### ISSUE TITLE
dhparam generates DSA parameters by default

### ISSUE THREAD

**Issue by kroeckx** [openssl/openssl#13148]
It seems that when we switched to the EVP interface in dhparam, we also switched to the deprecated dsa parameters when generating a DH key, with no way of still using a safe prime like the documentation suggest.

I currently don't see why we use the FIPS 186-4 algorithms for DH. It's not a requirement. They can also use one of the approved safe primes. They just can't use their own created safe prime in fips mode.

**Comment by mattcaswell** [comment]
It no longer does this in #13138

**Comment by kroeckx** [comment]
It still seems to be doing so:
```
Generating DH parameters, 2048 bit long safe prime
...............+.....................................+..+...+................................+.....+...............+.+......+.+..+.............+.+...+....................+............+....+.....+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++*
.............+........+...........+..+......+..........+.+.+....+...+.............................+..+........................................................+.....+..+.........+......+.......+..............+.....+.+..+.+............+.......+......+.......+....+................+....+........................+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++*
    DH Parameters: (2048 bit)
    P:   
        00:81:97:22:df:59:5e:c9:a4:fb:41:87:b0:20:fa:
        26:e6:40:b8:ac:a1:a5:f8:b0:cb:49:61:b3:63:02:
        7a:e2:ff:3b:ac:3d:5a:73:62:08:e3:a9:15:d1:ff:
        11:ec:9f:49:30:0a:c3:39:95:97:fa:03:0f:b9:35:
        25:b0:aa:24:11:76:dd:fd:09:62:81:d8:72:4c:61:
        47:45:01:5a:ed:24:97:94:c0:c4:b5:8b:f6:72:26:
        4f:5f:f8:79:08:fa:a6:fe:68:92:94:b1:d8:32:42:
        1b:a0:c2:c0:1f:db:1f:20:3a:ea:60:64:b6:b1:68:
        1c:48:4f:c4:58:19:18:de:08:5c:57:2c:db:22:c8:
        8b:87:7b:49:24:c3:d9:d0:b3:b4:36:72:fc:3f:f6:
        69:6b:71:29:2e:65:fd:1e:08:9e:64:ea:39:de:06:
        00:c1:84:f6:18:7e:70:54:f8:5e:3b:67:ba:3e:62:
        52:74:00:ef:c9:c6:bb:1e:b0:6f:13:64:a5:6f:8f:
        12:58:9e:be:95:4b:47:03:49:8c:9b:05:47:68:95:
        1d:4b:25:81:da:50:2f:2b:2f:40:2b:45:f3:5a:18:
        ec:d1:3d:3b:35:52:4d:39:a6:15:f1:fc:ea:41:31:
        ed:bc:f9:52:9b:ef:75:11:3e:ed:c7:80:be:13:14:
        b8:b7
    Q:   
        00:c3:59:54:8e:1a:2e:49:01:04:f9:ec:e0:21:20:
        fe:ea:25:f4:0d:ea:60:9c:b0:74:00:17:34:a9
    G:   
        5c:47:35:cc:5c:6d:96:2f:9b:db:d4:61:8f:6f:e1:
        30:91:70:71:ee:09:42:d2:88:80:6e:28:98:67:63:
        27:6e:b1:4b:a2:25:c4:ba:35:0f:d2:e3:83:98:c2:
        7f:d4:a4:8c:3b:a3:9c:6d:ab:4e:42:dc:db:69:3f:
        81:1d:8f:72:06:16:a3:3c:c7:13:cb:03:90:c1:2d:
        79:d4:29:4e:e6:40:f3:49:0c:34:b6:31:20:ef:55:
        fc:33:9a:57:25:27:04:37:9d:b7:d9:9d:b0:16:51:
        6b:9c:22:f7:2a:aa:8e:41:56:38:70:e7:3e:d1:04:
        62:7a:01:9f:0e:bd:81:80:34:6b:3d:e4:2d:78:a5:
        db:e4:78:d5:d1:41:08:44:c3:de:9a:cc:88:e3:a2:
        c0:71:fd:32:1b:f2:11:d8:69:ee:d1:8c:71:cf:61:
        01:0e:ea:56:b2:f2:b1:36:6a:6e:d6:f5:37:0b:10:
        b2:e5:67:9e:53:e7:14:b8:ae:15:97:83:7b:2a:c1:
        b6:3d:06:73:f3:3f:e0:e6:19:38:e7:e5:1a:c5:65:
        9e:46:e1:15:e7:a8:31:18:fe:70:5d:a9:10:c9:b2:
        6c:64:21:1d:2f:3a:b7:00:de:20:d8:cc:24:45:4c:
        5a:a7:13:c7:db:dc:11:a1:1d:9c:1a:16:8c:c0:19:
        18
    SEED:
        29:bc:3b:36:a4:75:28:28:9e:3f:c6:88:ba:52:f7:
        f2:c7:7d:40:b6:f5:bb:ac:fe:9f:08:45:df
    pcounter: 278
    h: 2
```

G is not 2, P does not seem like 2*Q+1.

**Comment by mattcaswell** [comment]
Hmmm. That doesn't look  right. Adding a "-out" parameter produces PKCS3 params...so:

$ openssl dhparam -out out.pem 2048
$ openssl -in out.pem -text -noout

Produced output without a "Q":

````
    DH Parameters: (2048 bit)
    P:   
        00:a9:e4:46:2b:1d:e4:93:c5:1f:5b:96:db:99:4d:
        e7:4d:32:42:e9:b6:87:3c:25:b6:75:f2:39:07:5a:
        bc:62:16:8b:44:0d:3f:0f:66:bc:94:7b:f5:f9:c6:
        10:2b:5d:a7:09:cd:61:cf:7a:84:aa:6a:4a:5b:81:
        fb:1d:6b:5a:dd:39:ed:0f:9f:38:19:f2:49:ed:8b:
        76:2f:8e:7a:2a:64:e2:84:88:d8:86:db:12:2f:fb:
        e3:cc:2b:7c:e0:ed:73:7d:a5:48:9c:ed:0d:42:4a:
        2c:26:db:77:f2:21:76:d7:99:ce:16:d7:e9:80:14:
        2a:fe:bc:14:37:a6:41:b7:87:ab:f3:29:88:48:e9:
        55:87:51:4a:36:4f:a7:b1:7f:cf:c1:49:81:4e:e2:
        11:8d:50:80:1f:c9:d0:53:05:26:b1:11:c0:e1:e6:
        06:b1:35:cf:ca:71:97:17:de:4d:27:47:71:f4:07:
        68:5e:06:27:75:b4:3e:87:18:33:c6:f6:6d:92:ef:
        f1:71:e7:cd:86:8c:f4:20:de:99:31:ff:e9:1e:34:
        c0:55:f1:33:b2:36:e5:14:ae:ad:a0:a8:d9:1f:53:
        36:2f:46:04:05:93:2b:25:1c:05:41:05:74:fb:ec:
        49:b4:c6:c5:1e:5d:a3:b6:49:a9:c1:87:46:43:79:
        fd:d5
    G:   
        40:e8:d0:6f:a7:df:6e:59:27:f6:02:e0:5c:9a:b0:
        7c:53:d2:fa:c4:7e:27:83:e3:d4:4c:de:fe:55:b5:
        26:53:c8:ab:9f:7e:2d:4f:77:5e:6b:2a:d3:e1:81:
        1d:b1:df:42:14:6a:38:71:4c:15:f7:b0:e1:c9:3a:
        2f:65:e8:80:69:5c:6b:d8:06:c8:3b:6d:54:de:22:
        29:4d:1a:ba:4c:6d:37:21:7e:74:2a:b7:ea:31:57:
        68:0b:08:a7:b8:a8:42:cb:a4:e8:6c:4e:e1:c7:eb:
        60:b8:fe:cf:47:c3:7d:72:ae:9b:ca:83:c5:7c:ef:
        99:4b:f1:a0:3e:72:5d:cb:cb:19:1f:ed:01:3f:db:
        d8:d6:25:88:ed:fd:09:8c:79:65:72:5d:0c:7c:c9:
        60:21:10:74:e6:a4:26:48:99:5f:c4:74:19:c2:af:
        12:cf:ba:ce:66:a3:b3:0d:37:d5:a1:fd:c6:ff:5d:
        38:c7:22:d3:7b:03:72:14:73:ab:4b:14:ca:e3:97:
        4a:58:e8:dc:42:eb:04:1e:ad:a3:a5:d3:43:60:b0:
        d1:06:d6:98:90:0c:6f:38:6c:4e:9f:7f:f3:77:e4:
        b0:aa:b6:99:e8:81:fe:33:f9:bc:4b:26:b0:15:1a:
        8b:fa:68:97:7a:a8:07:76:54:dc:ec:e7:bd:5c:d3:
        30
````
But it just seems to be DSA params with the Q dropped....

**Comment by slontis** [comment]
By DSA params you mean FFC parameters..

**Comment by mattcaswell** [comment]
This was fixed by #13228.
